File: 2016/05/5b.jl
using Nettle

tic()
input = "cxdnnyjw"
h = Hasher("md5")

function next_from(seed, n)
	while true
		update!(h, "$seed$n")
		hv = hexdigest!(h)
		n += 1
		if hv[1:5] == "00000"
			return hv, n
		end
	end
end

pass = Dict{Char,Char}()
n = 0
while length(pass) != 8
	hv, n = next_from(input, n)
	if hv[6] >= '0' && hv[6] <= '7' && !haskey(pass,hv[6])
		pass[hv[6]] = hv[7]
	end
end

toc()

File: 2016/05/src/bin/5b.rs
extern crate crypto;
use crypto::digest::Digest;
use crypto::md5::Md5;

struct PasswordGenerator<'a> {
    seed: &'a str,
    hasher: Md5,
}

impl<'a> PasswordGenerator<'a> {
    fn new(seed: &str) -> PasswordGenerator {
        PasswordGenerator {seed: seed, hasher: Md5::new()}
    }

    fn generate(&mut self) -> String {
       let mut n = 0;
       let mut password = vec![None; 8];
        loop {
            self.hasher.reset();
            self.hasher.input_str(format!("{}{}",self.seed,n).as_ref());
            n += 1;

            let rs =  self.hasher.result_str();
            if rs.chars().take(5).all(|c| c == '0') {
                let c = rs.chars().nth(5).unwrap();
                let pos = (c as u8 - 48) as usize;
                if c >= '0' && c <= '7' && password[pos].is_none() {
                    password[pos] = rs.chars().nth(6);
                }
            }
            if password.iter().all(|c| c.is_some()) {
                return password.iter().map(|c| c.unwrap()).collect();
            }
        }
    }
}

fn main() {
    let mut p = PasswordGenerator::new("cxdnnyjw");
    println!("The password is: {}", p.generate());
}

#[test]
fn test() {
    let mut p = PasswordGenerator::new("abc");
    assert_eq!(p.generate(), "05ace8e3".to_string());
}
